packages/runtime/container-runtime/src/opLifecycle/opGroupingManager.ts

```diff
@@ -96,19 +100,29 @@ export class OpGroupingManager {
 	}
 
 	public ungroupOp(op: ISequencedDocumentMessage): ISequencedDocumentMessage[] {
-		if (!isGroupContents(op.contents)) {
-			return [op];
+		assert(isGroupedBatch(op), "can only ungroup a grouped batch");
+
+		let messages: IGroupedMessage[] = [];
+		// ! Important to check "op.contents" first
+		if (isGroupContents(op.contents)) {
+			messages = op.contents.contents;
+		} else if (isGroupContents(op)) {
+			messages = op.contents;
 		}
 
-		const messages = op.contents.contents;
 		let fakeCsn = 1;
-		return messages.map((subMessage) => ({
-			...op,
-			clientSequenceNumber: fakeCsn++,
-			contents: subMessage.contents,
-			metadata: subMessage.metadata,
-			compression: subMessage.compression,
-		}));
+		return messages.map((subMessage) => {
+			// We don't want to preserve the "groupedBatch" type in case this ungrouped message is accidentally passed back through
+			const { type: _, ...props } = op;
```

Comment:
Took me a minute to recognize this syntax, for some reason the `_` variable name tripped me up.  Maybe `_dropped` instead?  Could be only me...

Reply:
Or do `type: undefined` in the `return` statement?